gramps_webapi/auth/oidc.py

```diff
@@ -0,0 +1,156 @@
+import uuid
+from authlib.integrations.flask_client import OAuth
+from flask import Blueprint, redirect, url_for, request, current_app, jsonify
+from flask_jwt_extended import create_access_token, create_refresh_token
+from sqlalchemy.exc import IntegrityError
+
+
+from . import user_db, User, add_user
+from .const import ROLE_USER
+from ..api.util import get_tree_id, abort_with_message, tree_exists
+from ..api.auth import get_permissions
+from ..const import TREE_MULTI
+
+
+oauth = OAuth()
+oidc_bp = Blueprint("oidc", __name__, url_prefix="/auth")
+
+
+def configure_oauth(app):
+    if not app.config.get("OAUTH_ENABLED", False):
+        return
+
+
+    oauth.init_app(app)
+
+
+    if app.config.get("OAUTH_GOOGLE_CLIENT_ID") and app.config.get("OAUTH_GOOGLE_CLIENT_SECRET"):
+        oauth.register(
+            name="google",
+            client_id=app.config["OAUTH_GOOGLE_CLIENT_ID"],
+            client_secret=app.config["OAUTH_GOOGLE_CLIENT_SECRET"],
+            access_token_url="https://oauth2.googleapis.com/token",
+            authorize_url="https://accounts.google.com/o/oauth2/auth",
+            api_base_url="https://www.googleapis.com/oauth2/v1/",
+            client_kwargs={"scope": "openid email profile"},
+        )
+
+
+    if app.config.get("OAUTH_GITHUB_CLIENT_ID") and app.config.get("OAUTH_GITHUB_CLIENT_SECRET"):
+        oauth.register(
+            name="github",
+            client_id=app.config["OAUTH_GITHUB_CLIENT_ID"],
+            client_secret=app.config["OAUTH_GITHUB_CLIENT_SECRET"],
+            access_token_url="https://github.com/login/oauth/access_token",
+            authorize_url="https://github.com/login/oauth/authorize",
+            api_base_url="https://api.github.com/",
+            client_kwargs={"scope": "read:user user:email"},
+        )
+
+
+    if app.config.get("OAUTH_MICROSOFT_CLIENT_ID") and app.config.get("OAUTH_MICROSOFT_CLIENT_SECRET"):
+        oauth.register(
+            name="microsoft",
+            client_id=app.config["OAUTH_MICROSOFT_CLIENT_ID"],
+            client_secret=app.config["OAUTH_MICROSOFT_CLIENT_SECRET"],
+            access_token_url="https://login.microsoftonline.com/common/oauth2/v2.0/token",
+            authorize_url="https://login.microsoftonline.com/common/oauth2/v2.0/authorize",
+            api_base_url="https://graph.microsoft.com/v1.0/",
+            client_kwargs={"scope": "openid email profile"},
+        )
+
+
+@oidc_bp.route("/login/<provider>")
+def login(provider):
```

Comment:
Doc string please

Reply:
And please add an example to the doc string, I see there is a `redirect_uri` argument expected in the request? What should it be?